core/src/Streamly/Internal/Data/Stream/StreamD/Transform.hs

```diff
@@ -442,35 +443,58 @@ postscanOnce (FL.Fold fstep initial extract) (Stream sstep state) =
             Stop -> return Stop
     step _ ScanDone = return Stop
 
-{-# INLINE scanOnce #-}
-scanOnce :: Monad m
-    => FL.Fold m a b -> Stream m a -> Stream m b
-scanOnce (FL.Fold fstep initial extract) (Stream sstep state) =
-    Stream step (ScanInit state)
+{-# INLINE scanWith #-}
+scanWith :: Monad m
+    => Bool -> Fold m a b -> Stream m a -> Stream m b
+scanWith restart (Fold fstep initial extract) (Stream sstep state) =
+    Stream step (ScanInit state, Nothing)
 
     where
 
     {-# INLINE_LATE step #-}
-    step _ (ScanInit st) = do
+    step _ (ScanInit st, _)= do
         res <- initial
         case res of
             FL.Partial fs -> do
                 !b <- extract fs
-                return $ Yield b $ ScanDo st fs
-            FL.Done b -> return $ Yield b ScanDone
-    step gst (ScanDo st fs) = do
+                return $ Yield b  (ScanDo st fs, Nothing)
+            FL.Done b -> return $ Yield b (ScanDone, Nothing)
+    step gst (ScanDo st fs, nxt) = do
         res <- sstep (adaptState gst) st
         case res of
             Yield x s -> do
                 r <- fstep fs x
                 case r of
                     FL.Partial fs1 -> do
                         !b <- extract fs1
-                        return $ Yield b $ ScanDo s fs1
-                    FL.Done b -> return $ Yield b ScanDone
-            Skip s -> return $ Skip $ ScanDo s fs
+                        return $ Yield b (ScanDo s fs1, Nothing)
+                    FL.Done b ->
+                        if restart
+                        then do
+                            if isNothing nxt
+                            then return $ Yield b (ScanDo s fs, Just b)
```

Comment:
I do not understand the point of keeping the second element of state tuple. This is unnecessarily complicated. All we need to do is unconditionally reinit the scan on done if the `restart` flag is true.